### PR TITLE
Enhance status for deploymentconfigs

### PR DESCRIPTION
**Pull request by kargakis**
Add various status fields for enhancing user experience. The fields are
also required for round-tripping upstream Deployments.

@ironcladlou @openshift/api-review  PTAL

Fixes https://github.com/openshift/origin/issues/5918
Part of https://trello.com/c/Mljldox7/643-5-deployments-downstream-support-for-upstream-features

@mfojtik @pweil- fyi

### CODE REVIEW

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 1]
```diff
@@ -137,7 +138,10 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		return fmt.Errorf("couldn't create deployment for deployment config %s: %v", deployutil.LabelForDeploymentConfig(config), err)
 	}
 	c.recorder.Eventf(config, "DeploymentCreated", "Created new deployment %q for version %d", created.Name, config.LatestVersion)
-	return nil
+	// Update deploymentConfig status; we probably need a separate call for status
```
Comment (deads2k) [inline pkg/deploy/controller/deploymentconfig/controller.go]: If we're introducing the idea of a `status`, let's introduce a `status` at least internally.  Since `status` should be read-only, we can be compatible by duplicating existing fields in two locations.

Does the deployer pod update logical status?  If so, then we'll need to need to respect changes on it if the `status` stanza is missing.  If not and its only the baked in controller, then we should be good to ignore status updates on old fields.

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: The deployer pod updates the status on the replication controller, it never touches the config. I think that `status` would be internal only, @ironcladlou thoughts?

Comment (deads2k) [inline pkg/deploy/controller/deploymentconfig/controller.go]: > The deployer pod updates the status on the replication controller, it never touches the config. I think that status would be internal only, @ironcladlou thoughts?

I'm fine with a deployer pod updating status.  I just wanted to make sure we handled it properly for backwards compatibility.  Sounds like we don't have to worry about it.

Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller.go]: The strategies just scale RCs up and down and don't touch the config. They also have no reason to read the new field via public API or anything. So I don't think introducing status to the config should have an effect outside the controllers.

File: pkg/deploy/api/types.go [hunk 2]
```diff
@@ -257,6 +257,8 @@ type DeploymentConfig struct {
 	// Details are the reasons for the update to this deployment config.
 	// This could be based on a change made by the user or caused by an automatic trigger
 	Details *DeploymentDetails
+	// Replicas is the total number of pods targeted by this deploymentConfig.
+	Replicas int
```
Comment (DirectXMan12) [inline pkg/deploy/api/types.go]: This name could get confusion -- we now have `DeploymentConfig.Replicas` and `DeploymentConfig.Template.Replicas`, one of which is read-only, and one of which now scales the DeploymentConfig, and it's not particularly clear which is which based on the name.  Since we don't have a status object here, perhaps this should be `TotalReplicas`, `CurrentReplicas`, or something of the sort, just to avoid confusion.

File: pkg/deploy/api/v1/types.go [hunk 3]
```diff
@@ -262,6 +262,8 @@ type DeploymentConfigStatus struct {
 	// Details are the reasons for the update to this deployment config.
 	// This could be based on a change made by the user or caused by an automatic trigger
 	Details *DeploymentDetails `json:"details,omitempty" description:"reasons for the last update to the config"`
+	// Replicas is the total number of pods targeted by this deploymentConfig.
+	Replicas int `json:"replicas,omitempty" description:"total number of pods targeted by this config"`
```
Comment (smarterclayton) [inline pkg/deploy/api/v1/types.go]: Is this upstream?

Comment (smarterclayton) [inline pkg/deploy/api/v1/types.go]: If so, you have API approval.

Comment (deads2k) [inline pkg/deploy/api/v1/types.go]: > Is this upstream?
> If so, you have API approval.

Upstream has a `status` object.

File: pkg/deploy/api/types.go (hunk 2 above)
Comment (ironcladlou) [inline pkg/deploy/api/types.go]: I think I missed why we can't bring the `Status` field internal.

File: pkg/deploy/api/v1/types.go (hunk 3 above)
Comment (kargakis) [inline pkg/deploy/api/v1/types.go]: Which status object has `replicas` for total replicas.

File: pkg/deploy/api/types.go (hunk 2 above)
Comment (kargakis) [inline pkg/deploy/api/types.go]: > I think I missed why we can't bring the Status field internal.

I am wondering about this. Our versioned dcs have status but not the internal one. Is there any reason why?

Comment (ironcladlou) [inline pkg/deploy/api/types.go]: The internal looks pretty weird because when we refactored the model for v1 (from which upstream is derived), we didn't refactor the internal model to match due to time constraints.

Comment (kargakis) [inline pkg/deploy/api/types.go]: OK, I think I want to change this here. Refactor the internal representation to use status and spec, add a status subresource and a separate call.

**Comment by kargakis** [comment]
Made this compile, can I have reviews especially for conversions (2nd commit) ? @deads2k @ironcladlou

**Comment by kargakis** [comment]
Actually hack/update-generated-conversions generates all conversions needed so I guess we don't need manual conversions any more... There are also other new conversions added that I am not sure of ...

**Comment by kargakis** [comment]
> There are also other new conversions added that I am not sure of ...

Hm, I think the rest are because of PodTemplateSpec...

File: pkg/deploy/api/types.go [hunk 4]
```diff
@@ -245,27 +245,46 @@ const DeploymentCancelledAnnotationValue = "true"
...
 	// DeploymentConfig with a new LatestVersion.
 	Triggers []DeploymentTriggerPolicy
-	// Template represents a desired deployment state and how to deploy it.
-	Template DeploymentTemplate
+
+	// Replicas is the number of desired replicas.
+	Replicas int
+
+	// Selector is a label query over pods that should match the Replicas count.
+	Selector map[string]string
```
Comment (deads2k) [inline pkg/deploy/api/types.go]: Is this new?  What was the existing field?

Comment (liggitt) [inline pkg/deploy/api/types.go]: can a single selector represent the replicas? What if you are rolling between two deployments which resulted from two versions of the config, and have different selectors?

Comment (deads2k) [inline pkg/deploy/api/types.go]: > can a single selector select all the replicas we care about? What if you are rolling between two deployments which resulted from two versions of the config, and have different selectors?

I asked @kargakis to split this into one pull for making spec/status with no other changes and then a separate pull to add the changes.  I *think* this is new, so that should help us find and see what this is doing.

Comment (kargakis) [inline pkg/deploy/api/types.go]: The new fields were behind DeploymentTemplate previously.

Comment (deads2k) [inline pkg/deploy/api/types.go]: > The new fields were behind DeploymentTemplate previously.

Ah, definitely interested in that separate pull then.  Checking congruence in one and then reasoning about changes in another..

Couldn't we have updated the `Selector` of the DC and orphaned existing pods for an old RC created by this DC?

File: pkg/deploy/api/types.go (hunk 2 above)
Comment (kargakis) [inline pkg/deploy/api/types.go]: Status is now a thing for our internal object.

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 5]
```diff
@@ -137,7 +138,10 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		return fmt.Errorf("couldn't create deployment for deployment config %s: %v", deployutil.LabelForDeploymentConfig(config), err)
 	}
 	c.recorder.Eventf(config, "DeploymentCreated", "Created new deployment %q for version %d", created.Name, config.Status.LatestVersion)
-	return nil
+	// Update deploymentConfig status; we probably need a separate call for status
```
Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Opened https://github.com/openshift/origin/issues/6465 for a separate status update call.

**Comment by kargakis** [comment]
[test]

File: pkg/deploy/registry/deployconfig/etcd/etcd.go [hunk 6]
```diff
@@ -152,48 +127,17 @@ func (r *ScaleREST) Update(ctx kapi.Context, obj runtime.Object) (runtime.Object
...
 
+	totalReplicas := deploymentConfig.Status.Replicas
 	oldReplicas := deploymentConfig.Spec.Replicas
 	deploymentConfig.Spec.Replicas = scale.Spec.Replicas
 	if err := r.registry.UpdateDeploymentConfig(ctx, deploymentConfig); err != nil {
 		return nil, false, err
 	}
+
+	scaleRet.Spec.Replicas = scale.Spec.Replicas
 	scaleRet.Status.Replicas = totalReplicas + (scale.Spec.Replicas - oldReplicas)
```
Comment (DirectXMan12) [inline pkg/deploy/registry/deployconfig/etcd/etcd.go]: With @ironcladlou's change to how DeploymentConfig scaling works, I don't think this calculation is accurate any more. We should just keep `TotalReplicas == oldReplicas`, since the replica count on RCs controlled by the DC won't change until the DC controller updates them.

**Comment by kargakis** [comment]
Flake in the deployment triggers integration test
```
I1222 13:09:46.141408   21420 controller.go:72] Waiting for first version of integration/config:0
I1222 13:09:46.145428   21420 controller.go:50] Created initial deployment for deploymentConfig "integration/config:0"
I1222 13:09:46.146497   21420 controller.go:61] Ignoring change for DeploymentConfig integration/config:1; no existing Deployment found
I1222 13:09:46.158297   21420 controller.go:74] Ignoring DeploymentConfig change for integration/config:1 (latestVersion=1); same as Deployment integration/config-1
--- FAIL: TestTriggers_configChange-4 (0.40s)
	deploy_trigger_test.go:319: Starting test openshift
	deploy_trigger_test.go:277: replicationControllers "config-1" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
FAIL
```
Hopefully will be fixed by https://github.com/openshift/origin/pull/6467

[test]

**Comment by liggitt** [comment]
Nope, the same issue has happened with in-process etcd

**Comment by kargakis** [comment]
https://github.com/openshift/origin/issues/6040 flake

**Comment by kargakis** [comment]
```
Error from server: 501: All the given peers are not reachable (failed to propose on members [https://172.18.4.185:4001] twice [last error: Unexpected HTTP status code]) [0]
!!! Error in hack/../test/end-to-end/core.sh:162
	'oc delete pod cli-with-token' exited with status 1
Call stack:
	1: hack/../test/end-to-end/core.sh:162 main(...)
Exiting with status 1
```

**Comment by kargakis** [comment]
[test]

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 7]
```diff
@@ -247,6 +254,11 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
 				glog.V(4).Infof("Updated deployment %q replica annotation to match current replica count %d", deployutil.LabelForDeployment(&deployment), newReplicaCount)
 			}
 		}
+		deployments = append(deployments, &deployment)
 	}
-	return nil
+	// TODO: Use a separate call for updating the status of a DC
+	// https://github.com/openshift/origin/issues/6465
+	config.Status.Replicas = kdeputil.GetReplicaCountForRCs(deployments)
+	_, err := c.osClient.DeploymentConfigs(config.Namespace).Update(config)
```
Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: @ironcladlou won't this result in a conflict if an older client had scaled the DC (L220)?

Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/controller.go]: I think you're right- shouldn't we be reassigning config local after the update on L220?

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 8]
```diff
@@ -137,7 +138,11 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
 		return fmt.Errorf("couldn't create deployment for deployment config %s: %v", deployutil.LabelForDeploymentConfig(config), err)
 	}
 	c.recorder.Eventf(config, "DeploymentCreated", "Created new deployment %q for version %d", created.Name, config.Status.LatestVersion)
-	return nil
+	// TODO: Use a separate call for updating the status of a DC
```
Comment (deads2k) [inline pkg/deploy/controller/deploymentconfig/controller.go]: We'll want an issue for this.

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: https://github.com/openshift/origin/issues/6465

File: pkg/deploy/controller/deploymentconfig/controller.go (hunk 7 above)
Comment (deads2k) [inline pkg/deploy/controller/deploymentconfig/controller.go]: > I think you're right- shouldn't we be reassigning config local after the update on L220?

Not finding the conflict in a test, suggests we have a test missing.

It seems like this update is more properly a patch until we have separate status updating.

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Would using another copy of the dc in the last update call not conflict?

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: > Not finding the conflict in a test, suggests we have a test missing.

We need to use older clients in our tests

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Or in this specific case we could emulate an older client by manually updating the oldest rc for a dc

**Comment by smarterclayton** [comment]
This API change is approved.

**Comment by kargakis** [comment]
I would say that this is blocked on https://github.com/openshift/origin/issues/6465 (so we can avoid resource conflicts in case of old client updates) which in turn is blocked on https://github.com/openshift/origin/issues/6337 (the trickiest one; planning to start working on it soon)

File: pkg/deploy/api/types.go [hunk 9]
```diff
@@ -314,14 +314,23 @@ type DeploymentConfigSpec struct {
 
 // DeploymentConfigStatus represents the current deployment state.
 type DeploymentConfigStatus struct {
-	// LatestVersion is used to determine whether the current deployment associated with a DeploymentConfig
-	// is out of sync.
+	// LatestVersion is used to determine whether the current deployment associated with a deployment
+	// cofnig is out of sync.
 	LatestVersion int
-	// Details are the reasons for the update to this deployment config.
-	// This could be based on a change made by the user or caused by an automatic trigger
```
Comment (stevekuznetsov) [inline pkg/deploy/api/types.go]: `s/cofnig/config/`

File: pkg/deploy/api/types.go [hunk 10]
```diff
@@ -314,14 +314,23 @@ type DeploymentConfigSpec struct {
...
+	// Details are the reasons for the update to this deployment config. This could be based on a
+	// change made by the user or caused by an automatic trigger.
 	Details *DeploymentDetails
-	// ObservedGeneration is the most recent generation observed by the controller.
+	// ObservedGeneration is the most recent generation observed by the deploymentconfig controller.
```
Comment (stevekuznetsov) [inline pkg/deploy/api/types.go]: Two words?

Comment (stevekuznetsov) [inline pkg/deploy/api/types.go]: Also I still think if we're not specially calling it a `DeploymentConfig` proper noun, since this is user facing doc, we should use "configuration" over "config".

File: pkg/deploy/api/v1/types.go [hunk 11]
```diff
@@ -278,14 +278,23 @@ type DeploymentConfigSpec struct {
 
 // DeploymentConfigStatus represents the current deployment state.
 type DeploymentConfigStatus struct {
-	// LatestVersion is used to determine whether the current deployment associated with a DeploymentConfig
-	// is out of sync.
+	// LatestVersion is used to determine whether the current deployment associated with a deployment
+	// cofnig is out of sync.
```
Comment (stevekuznetsov) [inline pkg/deploy/api/v1/types.go]: Same tytrip as above

**Comment by stevekuznetsov** [comment]
I missed the discussion that spurred this PR five months ago... Why do we want this? This seems like useful information to show a user, but the aggregation of the info from RCs, etc, could be done client side. Why add this into these structs?

**Comment by kargakis** [comment]
> I missed the discussion that spurred this PR five months ago... Why do we want this? This seems like useful information to show a user, but the aggregation of the info from RCs, etc, could be done client side. Why add this into these structs?

That's the job of a controller, to reconcile the spec and report on status. The dc controller already has a watch and knows all the required information for a given DC, why duplicate this work on clients? Also, we need the fields for parity with upstream Deployments.

**Comment by kargakis** [comment]
@mfojtik @pweil- this is part of https://trello.com/c/Mljldox7/643-5-deployments-downstream-support-for-upstream-features

**Comment by smarterclayton** [comment]
Each client should not have to recreate the world to know what is going
on.  A deployment config is a resource that abstracts many smaller resources

On May 12, 2016, at 6:52 PM, Michail Kargakis <notifications@github.com>
wrote:

I missed the discussion that spurred this PR five months ago... Why do we
want this? This seems like useful information to show a user, but the
aggregation of the info from RCs, etc, could be done client side. Why add
this into these structs?

That's the job of a controller, to reconcile the spec and report on status.
The dc controller already has a watch and knows all the required
information for a given DC, why duplicate this work on clients? Also, we
need the fields for parity with upstream Deployments.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/6233#issuecomment-218908275>

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 12]
```diff
@@ -263,17 +264,43 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
...
+			glog.V(2).Infof("Cannot update the status for %q: %v", deployutil.LabelForDeploymentConfig(config), err)
+			return transientError(fmt.Sprintf("cannot update the status for %q - requeuing", deployutil.LabelForDeploymentConfig(config)))
+		}
+		glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
+	}
 	glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
 	return nil
 }
+
+func calculateStatus(config deployapi.DeploymentConfig, deployments []kapi.ReplicationController) (*deployapi.DeploymentConfigStatus, error) {
```
Comment (stevekuznetsov) [inline pkg/deploy/controller/deploymentconfig/controller.go]: The actual config is only used for the error messages, so I think this method should really just return the three counts and not consume the config at all, or update the given config (and consume it as a pointer).

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 13]
```diff
@@ -263,17 +264,43 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
...
-	}
-	if _, err := c.osClient.DeploymentConfigs(config.Namespace).UpdateStatus(config); err != nil {
+func (c *DeploymentConfigController) updateStatus(config *deployapi.DeploymentConfig, deployments []kapi.ReplicationController) error {
+	newStatus, err := calculateStatus(*config, deployments)
+	if err != nil {
 		glog.V(2).Infof("Cannot update the status for %q: %v", deployutil.LabelForDeploymentConfig(config), err)
 		return transientError(fmt.Sprintf("cannot update the status for %q - requeuing", deployutil.LabelForDeploymentConfig(config)))
 	}
+
+	if !reflect.DeepEqual(*newStatus, config.Status) {
```
Comment (stevekuznetsov) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Simply updating the status would mean we don't need this (costly?) reflection call, as well.

File: pkg/deploy/util/util.go [hunk 14]
```diff
@@ -233,6 +235,66 @@ func MakeDeployment(config *deployapi.DeploymentConfig, codec runtime.Codec) (*a
...
+	totalReplicaCount := 0
+	for _, deployment := range deployments {
+		totalReplicaCount += deployment.Spec.Replicas
+	}
+	return totalReplicaCount
+}
+
+// GetActualReplicaCountForDeployments returns the sum of actual replicas for the
+// given deployments.
+func GetActualReplicaCountForDeployments(deployments []api.ReplicationController) int {
```
Comment (stevekuznetsov) [inline pkg/deploy/util/util.go]: "Actual" isn't making a meaningful distinction for me here - the above method returns the desired replicas, so perhaps "Actual" would mean more if that method was `GetDesiredReplicaCountForDeployments` our similar?

File: pkg/deploy/util/util.go [hunk 15]
```diff
@@ -233,6 +235,66 @@ func MakeDeployment(config *deployapi.DeploymentConfig, codec runtime.Codec) (*a
...
+	}
+	return readyPodCount
+}
+
+// GetPodsForDeployments returns all the pods for the provided deployments.
+func GetPodsForDeployments(c kclient.Interface, deployments []api.ReplicationController) ([]api.Pod, error) {
+	allPods := map[string]api.Pod{}
+	for _, deployment := range deployments {
+		options := api.ListOptions{LabelSelector: labels.Set(deployment.Spec.Selector).AsSelector()}
+		podList, err := c.Pods(deployment.ObjectMeta.Namespace).List(options)
```
Comment (stevekuznetsov) [inline pkg/deploy/util/util.go]: Won't this alias pods with the same name under different namespaces?

File: pkg/deploy/controller/deploymentconfig/controller.go [hunk 16]
```diff
@@ -263,17 +264,43 @@ func (c *DeploymentConfigController) reconcileDeployments(existingDeployments *k
...
+		}
+		glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
+	}
 	glog.V(4).Infof("Updated the status for %q (observed generation: %d)", deployutil.LabelForDeploymentConfig(config), config.Status.ObservedGeneration)
 	return nil
 }
+
+func calculateStatus(config deployapi.DeploymentConfig, deployments []kapi.ReplicationController) (*deployapi.DeploymentConfigStatus, error) {
+	// TODO: Implement MinReadySeconds for deploymentconfigs: https://github.com/openshift/origin/issues/7114
+	available, err := deployutil.GetAvailablePodsForDeployments(c.kubeClient, deployments, 0)
```
Comment (stevekuznetsov) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Name the int literal here for clarity

Comment (stevekuznetsov) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Would clarify the TODO as well

File: pkg/deploy/api/v1/types.go (hunk 11 above)
Comment (stevekuznetsov) [inline pkg/deploy/api/v1/types.go]: Typo, even.

File: pkg/deploy/controller/deploymentconfig/controller.go (hunk 13 above)
Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: This action will reflect its cost only on the present process while a redundant call to the apiserver may lead in update conflicts on other clients.

File: pkg/deploy/util/util.go (hunk 15 above)
Comment (kargakis) [inline pkg/deploy/util/util.go]: This helper is called for calculating the available pods for a deploymentconfig so all deployments and pods will be at the same namespace. I will add a comment clarifying it.

File: pkg/deploy/controller/deploymentconfig/controller.go (hunk 13 above)
Comment (stevekuznetsov) [inline pkg/deploy/controller/deploymentconfig/controller.go]: What call to the api server do you mean?

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: `c.osClient.DeploymentConfigs(config.Namespace).UpdateStatus(config)`

The controller is a client to the apiserver as every other client.

Comment (stevekuznetsov) [inline pkg/deploy/controller/deploymentconfig/controller.go]: I see. Fair enough, comments on the method remain !

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: This is WIP:) I will probably come back to it later today

File: pkg/deploy/controller/deploymentconfig/controller.go (hunk 12 above)
Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Not really. It is used for latestVersion, status details, generation, and you are going to use it for minReadySeconds.

Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: Also note that I am passing a copy and not a pointer.

File: pkg/deploy/controller/deploymentconfig/controller.go (hunk 16 above)
Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/controller.go]: ok

**Comment by kargakis** [comment]
@openshift/api-review can I get a review here?

**Comment by kargakis** [comment]
@smarterclayton I think these new fields are at the mercy of the resync interval of the deploymentconfig controller.

**Comment by kargakis** [comment]
I have identified two spots where status updates need to happen: one is after a deployment finishes (we could also have status updates from the deployer pod while it is scaling up/down) and one after scaling. The first case is covered by https://github.com/openshift/origin/pull/6233/commits/ae8c60b319a2ecddce183e9f4cefe1edf8eee361, but in the second case I am not sure what else is left except the resync interval. By the time etcd triggers a watch for the controller for a .spec.replicas change, most of the pods will not be ready, the dc will be reconciled and its status will be actually updated in the next resync.

**Comment by smarterclayton** [comment]
API approved.

**Comment by kargakis** [comment]
@mfojtik a review here can relax you from the rebase.

File: pkg/deploy/api/helpers.go [hunk 17]
```diff
@@ -32,8 +33,19 @@ func ScaleFromConfig(dc *DeploymentConfig) *extensions.Scale {
 		ObjectMeta: kapi.ObjectMeta{
 			Name:              dc.Name,
 			Namespace:         dc.Namespace,
+			UID:               dc.UID,
+			ResourceVersion:   dc.ResourceVersion,
 			CreationTimestamp: dc.CreationTimestamp,
 		},
+		Spec: extensions.ScaleSpec{
+			Replicas: int(dc.Spec.Replicas),
```
Comment (mfojtik) [inline pkg/deploy/api/helpers.go]: this will be changed to int32() in rebase I guess.

Comment (kargakis) [inline pkg/deploy/api/helpers.go]: Yeap. Waiting for it;)

File: pkg/deploy/controller/deploymentconfig/factory.go [hunk 18]
```diff
@@ -40,7 +40,7 @@ func (factory *DeploymentConfigControllerFactory) Create() controller.RunnableCo
 		},
 	}
 	queue := cache.NewFIFO(cache.MetaNamespaceKeyFunc)
-	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 2*time.Minute).Run()
+	cache.NewReflector(deploymentConfigLW, &deployapi.DeploymentConfig{}, queue, 30*time.Second).Run()
```
Comment (mfojtik) [inline pkg/deploy/controller/deploymentconfig/factory.go]: why this change?

**Comment by mfojtik** [comment]
@kargakis LGTM, nice job! but it does not really relaxed me :)

File: pkg/deploy/controller/deploymentconfig/factory.go (hunk 18 above)
Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/factory.go]: I still need @smarterclayton to approve this. It's the only issue with this PR. The upstream controller has the same sync interval so status for upstream Deployments will usually be updated faster than having it set to 2 minutes. See https://github.com/openshift/origin/pull/6233#issuecomment-219423116

Comment (smarterclayton) [inline pkg/deploy/controller/deploymentconfig/factory.go]: Hrm - I guess the next rebase (after this) we'll pull the list vs sync separation.  I'm concerned a bit that this will regress us a bit in performance, and whether we'll have to do a lot of work to pick that up.

@rrati do your changes to resync/relist depend on Informers, or can they work for any reflector?

Comment (rrati) [inline pkg/deploy/controller/deploymentconfig/factory.go]: That really depends on what is needed.:) The reflector itself will not exit on a resync notification any more.  It will break the watch, call Resync on the store object the reflector was configured to use, then initiate the watch.

However, the DeltaFIFO was the only store implementation that had access to the local cache and a notification means to allow the resync to cause the local cache to be inspected.  Other storage types will do nothing on a resync or just ensure that all items in the fifo are in the processing queue for the workers to process.

So, while the reflector will not exit on a resync call any more, I would say only the informer (or reflectors configured similarly to the informer) will actually cause an evaluation of the local cache.  Whether or not that is a big deal depends on what the use case is.

Comment (smarterclayton) [inline pkg/deploy/controller/deploymentconfig/factory.go]: Ok, so let's *not* change this for now, leave a TODO, and then someone
needs to own ensuring that the Origin controllers are taking advantage of
resync/relist correctly when it lands.

File: pkg/deploy/api/v1/types.go [hunk 19]
```diff
@@ -278,14 +278,23 @@ type DeploymentConfigSpec struct {
...
+	// change made by the user or caused by an automatic trigger.
 	Details *DeploymentDetails `json:"details,omitempty"`
-	// ObservedGeneration is the most recent generation observed by the controller.
+	// ObservedGeneration is the most recent generation observed by the deployment config controller.
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// Replicas is the total number of pods targeted by this deployment config.
+	Replicas int32 `json:"replicas,omitempty"`
+	// UpdatedReplicas is the total number of non-terminated pods targeted by this deployment config
+	// that have the desired template spec.
+	UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
```
Comment (ironcladlou) [inline pkg/deploy/api/v1/types.go]: Is it useful/important to know the updated available count?

Comment (kargakis) [inline pkg/deploy/api/v1/types.go]: Hm, we should discuss it upstream. For now, I think we are ok with AvailableReplicas.

**Comment by ironcladlou** [comment]
Overall design LGTM, needs tests now.

**Comment by kargakis** [comment]
The upstream controller [identifies updates on children (replicasets) and requeues the parent (deployment)](https://github.com/kubernetes/kubernetes/blob/62a8394eb491e82e5445e36542907608bf7b95f3/pkg/controller/deployment/deployment_controller.go#L242-L265). Doing that would most probably solve the issue with the resync we have here. We need to transition to using caches in the deployment controllers (right now we don't really use them).

**Comment by kargakis** [comment]
> The upstream controller identifies updates on children (replicasets) and requeues the parent (deployment). Doing that would most probably solve the issue with the resync we have here. We need to transition to using caches in the deployment controllers (right now we don't really use them).

@smarterclayton this means that we need to depart from the RetryController framework and start using informers. Shared caches will also be another thing. cc: @deads2k

**Comment by smarterclayton** [comment]
Yes.

File: pkg/deploy/controller/deploymentconfig/factory.go (hunk 18 above)
Comment (kargakis) [inline pkg/deploy/controller/deploymentconfig/factory.go]: Ok, got this working on top of https://github.com/openshift/origin/pull/9002 without the need to reduce the interval. Now all I need is the rebase to land so we can start using shared caches @mfojtik :)

**Comment by kargakis** [comment]
Rebased on top of the latest controller changes, shared a pod cache, added an extended test. This will need another look @mfojtik @ironcladlou @smarterclayton @deads2k

**Comment by kargakis** [comment]
[test]

File: pkg/deploy/api/v1/types.go (hunk 19 above)
Comment (smarterclayton) [inline pkg/deploy/api/v1/types.go]: Which of these are not upstream?

**Comment by mfojtik** [comment]
LGTM

File: pkg/deploy/api/v1/types.go (hunk 19 above)
Comment (kargakis) [inline pkg/deploy/api/v1/types.go]: Every field is upstream already

**Comment by kargakis** [comment]
[merge]

File: pkg/deploy/api/v1/types.go (hunk 19 above)
Comment (kargakis) [inline pkg/deploy/api/v1/types.go]: Except the one @ironcladlou proposes which is not in both

**Comment by kargakis** [comment]
This will also enable `oc rollout status`.

File: pkg/deploy/api/v1/types.go (hunk 19 above)
Comment (smarterclayton) [inline pkg/deploy/api/v1/types.go]: Ok, wanted to be clear.
